pkg/storageincentives/agent.go

```diff
@@ -369,6 +370,9 @@ func (a *Agent) handleClaim(ctx context.Context, round uint64) (bool, error) {
 }
 
 func (a *Agent) handleSample(ctx context.Context, round uint64) (bool, error) {
+
+	a.state.SetLastSelectedRound(round + 1)
```

Comment:
And as soon as we start handling a sample, we know our neighborhood has been selected, regardless of what other conditions may apply.  If a node sees Selected without Played, they know they missed out for some reason.

Reply:
And line 174 covers the sample that finishes calculating after the commit phase has started by re-triggering the commit event which gets us marked as Playing.

Reply:
@ldeffenb `handleSample` is called on begging on every `claim` phase. At this point agent still doesn't know if node is participating in this round until it calls `isParticipatingInUpcomingRound` in redistribution smart contract, right? If this is right, then logic in this line is not correct because `LastSelectedRound` is always going to be set. 

Imo, this should be corrected by rearranging early return conditions:

- check if `isParticipatingInUpcomingRound` first, early return on error or false
- set last selected round (at this point neighborhood is selected)
- check other conditions, and early return if those are meet 
- finally make sample

does this make sense ?

Reply:
Right, so the initial placement was correct? Just after the neighbourhood check?

Reply:
> Right, so the initial placement was correct? Just after the neighbourhood check?

Not quite exactly as initial. I am thinking that this should be way to go:

```
func (a *Agent) handleSample(ctx context.Context, round uint64) (bool, error) {
	storageRadius := a.store.StorageRadius()

	isPlaying, err := a.contract.IsPlaying(ctx, storageRadius)
	if err != nil {
		a.metrics.ErrCheckIsPlaying.Inc()
		return false, err
	}
	if !isPlaying {
		a.logger.Info("not playing in this round")
		return false, nil
	}

	a.metrics.NeighborhoodSelected.Inc()
	a.state.SetLastSelectedRound(round + 1)

	if !a.state.IsFullySynced() {
		a.logger.Info("skipping round because node is not fully synced")
		return false, nil
	}

	if a.state.IsFrozen() {
		a.logger.Info("skipping round because node is frozen")
		return false, nil
	}

	if !a.health.IsHealthy() {
		a.logger.Info("skipping round because node is unhealhy", "round", round)
		return false, nil
	}

	_, hasFunds, err := a.HasEnoughFundsToPlay(ctx)
	if err != nil {
		return false, fmt.Errorf("has enough funds to play: %w", err)
	}

	if !hasFunds {
		a.logger.Info("insufficient funds to participate in next round", "round", round)
		a.metrics.InsufficientFundsToPlay.Inc()
		return false, nil
	}

	sample, err := a.makeSample(ctx, storageRadius)
	if err != nil {
		return false, err
	}

	a.state.SetSampleData(round, sample)

	return true, nil
}
```

Reply:
Okay, so the neighbourhood check before the node state checks, right